### PR TITLE
`ParamsFrom` utility to determine types of params from Action Inputs

### PR DESCRIPTION
**Pull request by evantahler**
The new `ParamsFrom` TS utility can *finally* help dynamically determine the param types of an Action's run method!  This only works on Action classes that **statically** define the inputs, e.g. NOT in a constructor.  Here's a screenshot of things working:

Screen Shot 2021-12-20 at 7 14 01 PM

Note that `params.value` is properly a string, and we get an error that `params.foo` is undefined.

Here's the magic:
```ts
export type ParamsFrom<A extends Action | Task> = {
  [Input in keyof A["inputs"]]: A["inputs"][Input]["formatter"] extends (
    ...ags: any[]
  ) => any
    ? ReturnType<A["inputs"][Input]["formatter"]>
    : string;
};
```

What's nice about this approach is that we don't need to change any of Actionhero's internals, but we should start writing all of our Actions and Tasks with static definitions.

### CODE REVIEW

**Comment by evantahler** [comment]
Sadly, because the run method /implements/ the method from the abstract class, we cannot pre-determine the type or params within the ActionProcessor :(

File: src/classes/inputs.ts [hunk 1]
```diff
@@ -1,5 +1,15 @@
 import { Input } from "./input";
+import { Action } from "../classes/action";
+import { Task } from "../classes/task";
 
 export interface Inputs {
   [key: string]: Input;
 }
+
+export type ParamsFrom<A extends Action | Task> = {
```
Comment (gcoonrod) [inline src/classes/inputs.ts:9]: I don't love this syntax. Gives me flashbacks to Reflections in Java. I don't know TS well enough to critique this one way or the other though.

File: src/tasks/runAction.ts [hunk 2]
```diff
@@ -1,22 +1,18 @@
...
+  queue = "default";
 
-  async run(params: Record<string, any>) {
+  async run(params: ParamsFrom<RunAction>) {
     if (!params) params = {};
 
     const response = await action.run(
       params.action,
       params.version,
+      // @ts-ignore
```
Comment (gcoonrod) [inline src/tasks/runAction.ts:15]: Why is this needed?

**Review by gcoonrod — APPROVED** [review]
Overall I think this is a good idea. My TS knowledge isn't enough to speak authoritatively on the implementation though.

File: src/actions/cacheTest.ts [hunk 3]
```diff
@@ -1,39 +1,36 @@
...
+        value: "value",
+        expireTimestamp: 1420953274716,
+        createdAt: 1420953269716,
+        readAt: null as number,
       },
-    };
-  }
+      deleteResp: true,
+    },
+  };
```
Comment (krishnaglick) [inline src/actions/cacheTest.ts:33]: So we need to define these properties at the top-level in the class, rather than in the constructor to make it work? That seems sane.

File: src/classes/inputs.ts (hunk 1 above)
Comment (krishnaglick) [reply src/classes/inputs.ts:9]: Unfortunately it's the reality of Typescript that shit's going to get weird like this when you infer complex types. Theoretically `params: typeof this.params` should work but it doesn't.

File: src/tasks/runAction.ts (hunk 2 above)
Comment (krishnaglick) [reply src/tasks/runAction.ts:15]: Do Tasks need to be:
```
async run(params: {
    action: string;
    version: string;
    params: ParamsFrom<RunAction>;
  }) {
    const response = await action.run(
      params.action,
      params.version,
      params.params
    );
```
Maybe?

File: src/actions/cacheTest.ts (hunk 3 above)
Comment (evantahler) [reply src/actions/cacheTest.ts:33]: Yep.
